feat(mcdu): Progress Page Bearing/Dist To

tracernz opened the pull request:
<!-- ⚠⚠ Do not delete this pull request template! ⚠⚠ -->
<!-- Pull requests that do not follow this template are likely to be ignored. -->

<!-- Add the issues this PR fixes here. If no issues are related to this PR, then this line can be removed. -->
<!-- Add further issues with a full "Fixes #[issue_no]" line to ensure GitHub closes each one when the PR is merged. -->
<!--Fixes #[issue_no]-->

## Summary of Changes
<!-- Please provide a summary of changes for this pull request, ensuring all changes are explained. -->
I wrote this a while ago but never made a PR. Rebased and tidied up now so here it goes.
Implement bearing and distance to navaid/airport on the progress page. The values will be updated every 2 seconds when a navaid is entered (RSK 4). The duplicate names page is shown when the entered ident matches multiple navaids. ~~Currently only airports, waypoints, and navaids are supported. Lat/long, PBD, or PB/PB are not supported.~~ Supports all valid input formats: places (airports, runways, navaids), lat/long, place/bearing/distance, place-bearing/place-bearing.

Also a couple of small fixes to the duplicate names page:
 - If the distance is > 9999 Nm, show 9999, per the Honeywell step 1A manual
 - Correct the number of pages available at the top right

## Screenshots (if necessary)
<!-- If your PR includes visual changes, screenshots from before and after your change should always be included. -->
<!-- Please make best efforts to provide useful before and after screenshots. They should match camera angle, zoom, size, time of day, etc. -->
![image](https://user-images.githubusercontent.com/9995998/109407737-34320900-79e8-11eb-8816-de7e56cee8a4.png)
![image](https://user-images.githubusercontent.com/9995998/109407744-44e27f00-79e8-11eb-95aa-11636f8733e9.png)


## References
<!-- You should be making changes based on some kind reference (manuals, videos, IRL photos). P3D/xplane/fsx references will only be accepted if we believe that one cannot reasonably obtain a better source. Please post screenshots of the references you used. Ask around in the discord for how to find references for what you are working on. Exceptions will probably be made for IRL A320 pilots and engineers. -->

<!-- If you are making a pull request related to the MCDU, please make sure you are ONLY referencing the Honeywell Pegasus Step 1A (Rev 0), 2009 edition manual. -->
<!-- If you do not have this manual, please ask on our discord for assistance -->

Honeywell Step1A manual:
![image](https://user-images.githubusercontent.com/9995998/109407620-30ea4d80-79e7-11eb-8b78-5dbeeb1b942c.png)


## Additional context
<!-- Add any other context about the pull request here. -->

<!-- You may optionally provide your discord username, so that we may contact you directly about the issue. -->
Discord username (if different from GitHub):

## Testing instructions
<!-- Detail how this PR should be tested by QA. Try to list important items that need checking, either directly changed by this PR or that could be affected by it -->
- Enter a variety and waypoints, navaids, and airports; nearby and further away. Check the bearing and distance is displayed correctly, and is reasonable
- Enter a navaid approx. 10-20 nm away, and fly the bearing shown. You should end up flying directly inbound to the navaid
- Enter a navid that has a duplicate ident somewhere in the world (especially 2/3 digit ones). The duplicate names page should allow you to select the correct navaid.
- Enter a runway e.g. NZWN34 or NZAA05R while sitting at the threshold, check distance is 0 or close to
- Enter the opposite end of the runway and ensure bearing is runway heading, dist is approx. runway length
- Try entering place/bearing/distance
- Try entering place-bearing/place-bearing (suggest a runway if you're struggling, e.g. NZWN16-120/NZWN34-30 for a point to the east of NZWN 16/34)
- Try entering the various formats with places/navaids that don't exist and make sure you get an appropriate scratchpad message

<!-- DO NOT DELETE THIS -->
## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

beheh commented:
While I haven't got a Honeywell-specific reference I know that at least on Thales you're actually able to enter specific airport/runway combination in that field, e.g. `EGLL27R` as an example. I would assume that that's also a Honeywell-feature ("can be any [...] runway"):

![image](https://user-images.githubusercontent.com/929769/109414930-19df3780-79b6-11eb-82ce-4f9720e88711.png)

krangerich commented:
Fixes #1744

danbw9 commented:
I can confirm with the above that it is possible with Honeywell to enter a runway identifier eg. EKCH22L. This is what is usually entered for situational awareness for departure and arrival.

MisterChocker commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_ProgressPage.js:
```diff
@@ -89,7 +104,7 @@ class CDUProgressPage {
             ["{small}*{end}[\xa0\xa0\xa0\xa0][color]cyan"],
             ["\xa0\xa0BRG / DIST"],
-            ["{small}\xa0---°/----.-{end}", "{small}TO{end} {cyan}[{sp}{sp}{sp}{sp}{sp}]{end}"],
+            [progBearingDist, progWaypoint, "{small}\xa0TO{end}"],
```
When using the center position, the text alignment is totally off.

```suggestion
            [progBearingDist + "{small}\xa0TO{end}", progWaypoint],
```

tracernz replied:
So this was based on the Honeywell manual and some guesswork... now I've found a couple of [real](https://youtu.be/efQlfF87c8c?t=225) [references](https://youtu.be/twwUjHXdNVU?t=760) and I don't think either my version or yours is correct. I'm going to mark this a draft for now and make this match IRL, and also look into feasibility of supporting runways as requested above.

MisterChocker commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js:
```diff
@@ -301,6 +302,8 @@ class FMCMainDisplay extends BaseAirliners {
 
         this.updateDisplayedConstraints();
+
+        this.updateProgDistance(_deltaTime);
```
You might want to use an update throttler or call this method when opening/loading the prog page.
The prog page will be refreshed every x times a second anyways.

tracernz replied:
> update throttler

I'll look into that.

beheh commented:
Is anything still missing here? What's missing to get this ready to review?

tracernz commented on A32NX/html_ui/JS/A32NX_Util.js:
```diff
@@ -46,6 +46,59 @@ A32NX_Util.createMachine = (machineDef) => {
+
+/**
+ * Computes the intersection point of two bearings on a great circle
+ * @param {LatLong | LatLongAlt} latlon1
```
Is this correct way for a union of multiple types?

beheh replied:
Yes: https://jsdoc.app/tags-param.html#multiple-types-and-repeatable-parameters

tracernz commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js:
```diff
@@ -3120,6 +3126,246 @@ class FMCMainDisplay extends BaseAirliners {
+                        if (Avionics.Utils.formatRunway(airport.infos.oneWayRunways[i].designation) === rwy[2]) {
+                            const runway = airport.infos.oneWayRunways[i];
+                            // this should be to the treshold but we don't have that, so we just use half the length
+                            const adjustedCoordinates = Avionics.Utils.bearingDistanceToCoordinates((runway.direction + 180) % 360, runway.length / 2 * 0.000539957, runway.latitude, runway.longitude); // TODO proper m to nm conversion
```
Do we have unit conversions somewhere? I haven't found anything but I think it would be useful to have a little util class somewhere. Maybe I can add it if it doesn't exist.

tracernz commented:
> 
> 
> Is anything still missing here? What's missing to get this ready to review?

I have added a couple of coments here. I'd appreciate if you can offer your thoughts.

tracernz commented on A32NX/html_ui/Pages/A32NX_Utils/NXUnits.js:
```diff
@@ -0,0 +1,49 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+let NXUnits = {};
```
@beheh I bet you can help me come up with a much better implementation.. 😉

tracernz commented:
I punted on adding a unit conversion API for my one use in this PR (can come later). This is now ready for review.

tracernz commented:
Fixed some lint errors 😅

marcman86 commented:
Category: QA Tester
Name: marcman86#4907
Date of testing: 28.3.2021
Version of the sim: 1.14.6.0
PR Tested: #3752
Tier of testing: 2
Changes to observe: PERF/W&B and fuel prediction page inputs
Testing technique: Tryout

Steps made to test the Pull Request:

1.) Spawned on runway at EDDP
2.) Took off to 5000ft and flight around
3.) Tryed PROG page Bearing/Dist

- Airport
- Waypoint
- VOR
- ADF
- Runway

Issues: 
- RETURN on DUPLICATE NAMES pages don't bring back the PROG page. It goes to F-PLN page.
- On VOR/ADF flyover there is no deadzone ("signal loss")

Overall Rating: Mostly good

Conclusions of the testing: Very nice PR! Just a minor issues.

Hope this helps :-)

MisterChocker commented:
> * RETURN on DUPLICATE NAMES pages don't bring back the PROG page. It goes to F-PLN page.

This one should be fixed (if complies with fcom)

tracernz commented on A32NX/html_ui/JS/A32NX_Util.js:
```diff
@@ -46,6 +46,59 @@ A32NX_Util.createMachine = (machineDef) => {
+    // c.f. https://blog.mbedded.ninja/mathematics/geometry/spherical-geometry/finding-the-intersection-of-two-arcs-that-lie-on-a-sphere/
+
+    // TODO should be magvar at the lat, lon
+    const magVar = SimVar.GetSimVarValue("MAGVAR", "degree");
```
I had forgotten about this. I think this is the wrong level of abstraction to be dealing with magnetic headings. [Magvar looks to available on navaids from the facility loader since SU3](https://github.com/flybywiresim/a32nx/pull/3868/files#diff-1a4f0f0dbbd73af4b8e2bea0c691cfc053b7d51808c4f3017d31003f39f8db35R298) so I can get the proper mag var now too. I will refactor this a little in addition to the other note from QA.

tracernz commented:
> On VOR/ADF flyover there is no deadzone ("signal loss")

That's an interesting point. If it is computed from the data in the nav database there wouldn't be.

beheh commented:
There's the cone of confusion, but our RADNAV generally does not implement this yet (as the simulated signals are perfect and don't have this issue).

tshomas commented:
If this Cone of Confusion stuff is still talking about the bearing/dist page, doesn't apply as it's based off the nav database. So you can set a VOR on the page on the other side of the world and it still will work, and flying overhead you won't lose signal.

marcman86 commented:
What i ment is like on the ND needles. The VOR disapears for a moment on flyover. But i think this is due to simulate the "radio signal loss" above the antenna. I think in the bearing/dist all data comes from GPS positions and is just a comparison between its and the AC position.

tracernz commented:
Rebased, and added two new commits (the last two). Please check you're still happy. :)

tracernz commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_SelectWptPage.js:
```diff
@@ -55,7 +55,7 @@ class A320_Neo_CDU_SelectWptPage {
                 };
                 mcdu.onLeftInput[5] = () => {
-                    CDUFlightPlanPage.ShowPage(mcdu, 0);
+                    mcdu.returnCallback(mcdu);
```
This is the simplest way I can think of that doesn't involved passing a returnPage callback through a chain of functions.

easyjetsimpilot commented:
Category: QA Tester
Name: Gavin (easyjetsimpilot)
Date of testing: 01.4.2021
Version of the sim: 1.14.6.0
PR Tested: #3752
Tier of testing: 2
Changes to observe: Progress Page Bearing/Dist To
Testing technique: Launched ready to fly at LEPA 06L runway threshold (payware scenery)

Steps made to test the Pull Request:

1) Spawned on runway at LEPA06L checked distance on PROG page shown to be LEPA06L - 0.2NM
2. LEPA24R showed correct distance and bearing to opposite end of the runway
3) Checked various waypoints and VORs near and far away & concluded accurate distance and headings shown
4) Entered a navaid 20 miles from LEPA and flew track heading shown. Passed directly over it - excellent!
5) Entered MCT VOR and brought up duplicate name pages to select from
6) Tested place/bearing/distance (EGCC/090/30) and displayed correct information 
7) Entered incorrect format and obtained the FORMAT ERROR
8) Entered made up navaid and shown the NOT IN DATABASE message.


Conclusions of the testing: Excellent PR - one that will be extremely valuable as is in the real world and used constantly to assist with TOD calculations. Can't wait for this to be merged!

Benjozork commented:
@tracernz needs to be rebased

tracernz commented:
> @tracernz needs to be rebased

git has really improved in this aspect in the last few years! That went without a hitch despite the rename. 😅

MisterChocker reviewed (approved):
No, my requested changes have been dealt with to my satisfaction.

ExampleWasTaken commented:
Category: QA Trial
Name: `ExampleWasTaken#0886`
Date of testing: 08/04/2021
Version of the sim: 1.14.6.0
PR Tested: #3752 
Tier of testing: 1
Changes to observe:
Progress Page Bearing/Dist To

Testing technique:
1. Entered a navaid approx. 10-20 nm away, and flew the bearing shown. Ended up flying directly inbound to the navaid
2. Entered a navid that has a duplicate ident. The duplicate names page allowed to select the correct navaid.
3. Entered a runway e.g. NZWN34 or NZAA05R while sitting at the threshold, checked distance is 0 or close to
4. Entered the opposite end of the runway and ensured bearing is runway heading, dist is approx. runway length
5. Entered place/bearing/distance
6. Try entering place-bearing/place-bearing
7. Entered the various formats with places/navaids that don't exist - got appropriate scratchpad message

Overall Rating: LGTM
Conclusions of the testing: Very happy to see that addition. Will be nice for flights without a flight plan loaded.

FlightSimmer7700 commented:
Category: QA Tester
Name: FlightSimmer7700#3468
Date of testing: 8/4/21
Version of the sim: 1.14.6.0
PR Tested: #3752 
Tier of testing: 2
Changes to observe:
Bearing/distance to

Testing technique: Did a full flight
Steps you've made to test the Pull Request:
1. Loaded in
2. Took off
3. put waypoints on and off route into the progress page
4. put runways
5. airports/waypoints far and close
6. Note: haven't tested :Try entering place-bearing/place-bearing (suggest a runway if you're struggling, e.g. NZWN16-120/NZWN34-30 for a point to the east of NZWN 16/34)

Media: Insert screenshots, recordings, or anything relevant to the review here.
![image](https://user-images.githubusercontent.com/70103461/113949390-10dc5400-9807-11eb-976d-b23423e5b38b.png)

Issues (If Any): As shown by the picture above the nd and the mcdu distances are different and if possible it would be helpful if these two updates at the same time as they can sometimes be quite a bit off and its random which one is first, discussion on discord: https://discord.com/channels/738864299392630914/754130199804772372/829494206673453126


Overall Rating: Good
Conclusions of the testing: Everything works as expected apart from the distances being different which is why I have left a rating of good because its more of a request really

tracernz commented:
> Conclusions of the testing: Everything works as expected apart from the distances being different which is why I have left a rating of good because its more of a request really

The real MCDU only updates the bearing/distance every 2 seconds, so I have matched that here.

![image](https://user-images.githubusercontent.com/9995998/113951460-b9ba9b80-9867-11eb-8eae-111b5caa3e37.png)

tracernz commented:
Found a bug, but heading out the door. Will fix later.

St54Kevin commented:
Category: QA Tester
Name: St54Kevin#3583
Date of testing: 08/04/2021
Version of the sim: 1.14.6.0
PR Tested: #3752 
Tier of testing: 1
Changes to observe: Bearing/Dist To

**Steps you've made to test the Pull Request:**
1. Loaded onto RW
2. Checked VOR, NDB, Airport, RW and Waypoints
3. Quick patterns to see behavior

**Media:**

**Issues (If Any):** - 

**Conclusions of the testing:** LGTM. Thanks this is a very helpful feature!

tracernz commented:
Here's how I'm testing the PB/PB:
Looking at the SkyVector chart (actually I used Navigraph...), we see NZWN near the bottom, NZPP at the west side of the island, and MAMOD waypoint out to the east. There is an airway Q825 from NZWN to MAMOD at a bearing of 47° and distance of 32 NM, and airway Q546 from NZPP to MAMOD at a bearing of 101°. I spawn at the gate at NZWN, power up, and align ADIRS. I enter NZWN-47/NZPP-101 into the BRG/DIST field. It measures 047°/31.5. I then enter MAMOD directly, and it measures 048°/31.7. Given these bearings are all only accurate to the nearest degree I'm happy.

https://skyvector.com/?ll=-40.97416112017399,175.08828734098648&chart=302&zoom=1

I've rebased. fixed the dumb bug I introduced with my recent refactor, which affected bearings for everything except PBD. Tests okay at my end.

marcman86 commented:
Category: QA Tester
Name: marcman86#4907
Date of testing: 8.04.2021
Version of the sim: 1.14.6.0
PR Tested: #3752
Tier of testing: 1
Changes to observe: PROG Page Bearing/Dist To

Testing technique: Tryout

Steps made to test the Pull Request:

1.) Spawned on runway at EDDP
2.) Took off to 5000ft and flight around
3.) Tryed PROG page Bearing/Dist

- Airport
- Waypoint
- VOR
- ADF
- Runway

Return to PROG works now.

Issues: -

Overall Rating: Good

Conclusions of the testing: Works

Hope this helps :-)

BritishAvgeek commented:
Category: QA Tester
Name: BritishAvgeek#1436
Date of testing: 12/04/2021
Version of the sim: 1.14.6.0
PR Tested: #3752
Tier of testing: 2
Changes to observe: Progress Page Bearing/Dist To

Testing technique: Full flight Manchester to Milan on VATSIM

Steps you've made to test the Pull Request:
1. Spawned Cold and Dark and Gatwick, set up for flight to Milan from EGCC
2. Departed and during the SID, set up a progress page bearing/dist to DVR (dover). It displayed it perfectly well
3. Continued to test with numerous waypoints further away and closer to the aircraft throughout the journey
4. Input a progress input for dist/bearing to ODINA and monitored that for 30 minutes as we progressed towards it, updated throughout and everything worked as expected in the Prog page.

Flight notes (When done a full flight):
Route: EGCC SANBA1R SANBA UN859 HON L10 RINTI UL10 LESDO UL15 BEGAR L15 RIPUS N850 ODINA ODIN2G LIMC

Media: Livestreamed here https://youtu.be/SbbJH42vS-w

![image](https://user-images.githubusercontent.com/71381591/114397175-4af67e80-9b96-11eb-8eb5-96df2ec9efe4.png)
![image](https://user-images.githubusercontent.com/71381591/114397244-5fd31200-9b96-11eb-9a4d-16677a903574.png)
![image](https://user-images.githubusercontent.com/71381591/114398580-db818e80-9b97-11eb-9792-d6a5acb26ca7.png)

Random Airport not linked to the route but nearby to check the function off-plan:
![image](https://user-images.githubusercontent.com/71381591/114398617-e76d5080-9b97-11eb-9c48-e97403b32fa6.png)

Issues (If Any): Had issues NOT linked to this PR - direct to wiping the flight plan once and direct to wiping the entire STAR just prior to entry as well which wasn't ideal... but not linked to this PR.

Overall Rating: Good
Conclusions of the testing: Had no issues with this PR, PROG page worked as expected/described... thanks!

beheh commented:
In the video from @BritishAvgeek above, [at one point](https://www.youtube.com/watch?v=SbbJH42vS-w#t=32m46s) entering a waypoint name brings up a duplicate names page with exactly one waypoint:

![image](https://user-images.githubusercontent.com/929769/114413691-813ff800-9bae-11eb-81a5-f5b4604717d5.png)

I don't think this should ever happen.

tracernz commented:
> In the video from @BritishAvgeek above, [at one point](https://www.youtube.com/watch?v=SbbJH42vS-w#t=32m46s) entering a waypoint name brings up a duplicate names page with exactly one waypoint:

Quite strange https://github.com/flybywiresim/a32nx/blob/5ad17d128c66a85e011263e32d5ec704896881a2/flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/FMC/A32NX_FMCMainDisplay.js#L1659
I guess the page either has filtering that should be in FMC instead, or it has a bug. I will look later. If it's a simple fix I'm happy to do it here, but otherwise it should be an issue in it's own right rather than further holding this up (as the bug already exists outside of this PR).

tracernz commented:
I am going to open another PR to fix the duplicate waypoint page as it affects a few other places and will need it's own QA (and the bug exists outside of this PR).

TheSledge-SLEJ commented:
A thing of great beauty
